File: 0020-Valid-Parentheses/lc0020.cc
// C++ solution.

class Solution {
public:
    inline int findRType(char par) {
        if (par == ')') { return 0; }
        if (par == '}') { return 1; }
        if (par == ']') { return 2; }
        return -1;
    }

    bool isValid(string s) {
        stack<char> stk;
        for (char c : s) {
            int rtype = findRType(c);
            if (rtype >= 0) {
                if (stk.empty()) { return false; }
                if (stk.top() != lpar[rtype]) { return false; }
                stk.pop();
            } else {
                stk.push(c);
            }
        }
        return stk.empty();
    }

private:
    const array<char, 3> lpar{'(', '{', '['};
    const array<char, 3> rpar{')', '}', ']'};
};

File: 0020-Valid-Parentheses/lc0020.rs
// Rust solution.

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut st: Vec<char> = Vec::new();
        let rpar: [char; 3] = [')', '}', ']'];
        let lpar: [char; 3] = ['(', '{', '['];

        for c in s.chars() {
            let mut rmatched = false;
            for i in 0..3 {
                if c == rpar[i] {
                    match st.pop() {
                        Some(top) => {
                            if top != lpar[i] {
                                return false;
                            } else {
                                rmatched = true;
                            }
                        },
                        None => { return false; }
                    }
                }
            }
            if !rmatched { st.push(c); }
        }
        st.is_empty()
    }
}
